python/mspasspy/db/database.py

```diff
@@ -308,53 +375,726 @@ def save_data(self, mspass_object, storage_mode='gridfs', dfile=None, dir=None,
+            else:
+                # attribute values remain the same
+                update_dict[unique_k] = doc[k]
+                
+        # update the fixed attributes in the document in the collection
+        filter_ = {'_id': doc['_id']}
+        # use replace_one here because there may be some aliases in the document
+        col.replace_one(filter_, update_dict)
+        
+        return fixed_cnt, print_messages
+
+    def verify(self, document_id, collection='wf', tests=['xref', 'type', 'undefined']):
+        """
+        Verify a document in a collection, including checking links, checking required attributes and checking if attribute type matches schema.
+
+        :param document_id: the value of the _id field in the document you want to verify
+        :type document_id: class:`bson.objectid.ObjectId`
+        :param collection: the name of collection saving the document. If not specified, use the default wf collection
+        :param tests: the type of tests you want to verify, should be a subset of ['xref', 'type', 'undefined']
+        :type tests: :class:`list` of :class:`str`
+ 
+        :return fixed_cnt: number of keys that are fixed
+        :rtype fixed_cnt: class:`dict`
+        :return print_messages: the verbose informative/invalid messages
+        :rtype print_messages: class:`list`
+        """
+        # check tests
+        for test in tests:
+            if test not in ['xref', 'type', 'undefined']:
+                raise MsPASSError('only xref, type and undefined are supported, but {} is requested.'.format(test), 'Fatal')
+        # remove redundant if happens
+        tests = list(set(tests))
+
+        problematic_keys = {}
+
+        collection = self.database_schema.default_name(collection)
+        col = self[collection]
+        doc = col.find_one({'_id': document_id})
+
+        # if the document does not exist in the db collection, return
+        if not doc:
+            return problematic_keys
+        
+        # run the tests
+        for test in tests:
+            if test == 'xref':
+                # test every possible xref keys in the doc
+                for key in doc:
+                    if self.database_schema[collection].is_defined(key) and self.database_schema[collection].is_xref_key(key):
+                        # only test normalized keys with '_id'
+                        if '_id' in key and key.rsplit('_', 1)[1] == 'id':
+                            normalized_collection_name = key.rsplit('_', 1)[0]
+                            normalized_collection_name = self.database_schema.default_name(normalized_collection_name)
+                            normalized_col = self[normalized_collection_name]
+                            # try to find the referenced docuement
+                            normalized_doc = normalized_col.find_one({'_id': doc[key]})
+                            if not normalized_doc:
+                                problematic_keys[key] = test
```

Comment:
It seems we can use these in the `check_links` method instead of what's there right now. It seems the same argument applies to the undefined test and the type test below.